word_language_model/model.py

```diff
@@ -55,3 +57,165 @@ def init_hidden(self, bsz):
                     weight.new_zeros(self.nlayers, bsz, self.nhid))
         else:
             return weight.new_zeros(self.nlayers, bsz, self.nhid)
+
+class TransformerSeq2Seq(nn.Module):
+    r"""A transformer model applied for sequence-to-sequence transform.
+        User is able to modified the attributes as needed.
+    Args:
+        src_vocab: the number of vocabularies in the source sequence (required).
+        tgt_vocab: the number of vocabularies in the target sequence (required).
+        d_model: the dimension of the encoder/decoder embedding models (default=512).
+        nhead: the number of heads in the multiheadattention models (default=8).
+        num_encoder_layers: the number of sub-encoder-layers in the encoder (default=6).
+        num_decoder_layers: the number of sub-decoder-layers in the decoder (default=6).
+        dim_feedforward: the dimension of the feedforward network model (default=2048).
+        dropout: the dropout value (default=0.1).
+    Examples::
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab)
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab, nhead=16, num_encoder_layers=12)
+    """
+
+    def __init__(self, src_vocab, tgt_vocab, d_model=512, nhead=8, num_encoder_layers=6,
+                 num_decoder_layers=6, dim_feedforward=2048, dropout=0.1):
+        super(TransformerSeq2Seq, self).__init__()
+        self.transformer = nn.Transformer(d_model=d_model, nhead=nhead,
+                                          num_encoder_layers=num_encoder_layers,
+                                          num_decoder_layers=num_decoder_layers,
+                                          dim_feedforward=dim_feedforward, dropout=dropout)
+        self.src_embed = nn.Embedding(src_vocab, d_model)
+        self.pos_encoder = PositionalEncoding(d_model, dropout)
+        self.tgt_embed = nn.Embedding(tgt_vocab, d_model)
+        self.pos_decoder = PositionalEncoding(d_model, dropout)
+
+        self.generator = Generator(d_model, tgt_vocab)
+        self.src_vocab = src_vocab
+        self.tgt_vocab = tgt_vocab
+        self.d_model = d_model
+        self.nhead = nhead
+        self.num_encoder_layers = num_encoder_layers
+        self.num_decoder_layers = num_decoder_layers
+        self.dim_feedforward = dim_feedforward
+        self.dropout = dropout
+
+        self._reset_parameters()
+
+    def _reset_parameters(self):
+        r"""Initiate parameters in the model."""
+
+        for p in self.parameters():
+            if p.dim() > 1:
+                nn.init.xavier_uniform_(p)
+
+    def forward(self, src, tgt, src_mask=None, tgt_mask=None, memory_mask=None):
```

Comment:
Can these masks be cached via member variables?

Reply:
If those attributes are not used inside the transformer model, I will delete them to release the memory.

Reply:
The masks depends on the incoming src/tgt sequences.

Reply:
But it looks like they're regenerated every time if called outside of this as well. So then let's at least do that in here to hide it I'd say.